### PR TITLE
TINY-10916: Transparent color is converted into #00000

### PR DESCRIPTION
**Pull request by ArvinJ-H**
Related Ticket: 
TINY-10916

Description of Changes:
- Added a seperate if statement to cover the edge case of transparent

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/`, `hotfix/` or `spike/`

Review:
* [x] Milestone set
* [x] <s>Docs ticket created (if applicable)</s>

GitHub issues (if applicable):
https://github.com/tinymce/tinymce/issues/9391

### CODE REVIEW

File: modules/tinymce/src/core/main/ts/api/html/Styles.ts [hunk 1]
```diff
@@ -246,6 +247,9 @@ const Styles = (settings: StylesSettings = {}, schema?: Schema): Styles => {
             name = decodeHexSequences(name);
             value = decodeHexSequences(value);
 
+            console.log('name', name);
+            console.log('value', value);
```
Comment (spocke) [inline modules/tinymce/src/core/main/ts/api/html/Styles.ts:251]: Remove these.

```suggestion
            console.log('name', name);
            console.log('value', value);
```

File: .changes/unreleased/tinymce-TINY-10916-2024-05-15.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+project: tinymce
+kind: Fixed
+body: Fixed a bug where applying a transparent color resulted in it being converted to black.
```
Comment (spocke) [inline .changes/unreleased/tinymce-TINY-10916-2024-05-15.yaml:3]: It converts existing colors as well so just loading up a document with `background: transparent` will change the color.

File: modules/tinymce/src/core/test/ts/browser/html/StylesTest.ts [hunk 3]
```diff
@@ -264,4 +264,9 @@ describe('browser.tinymce.core.html.StylesTest', () => {
     assertStyles(styles, 'color: rgba(1, 2, 3);', 'color: rgba(1, 2, 3);');
     assertStyles(styles, 'color: rgb(1, 2, 3, 0.5);', 'color: rgb(1, 2, 3, 0.5);');
   });
+
+  it('TINY-10916: transparent should be convert to black with 0 alpha', () => {
+    const styles = Styles();
+    assertStyles(styles, 'color: transparent;', 'color: #00000000;');
```
Comment (spocke) [inline modules/tinymce/src/core/test/ts/browser/html/StylesTest.ts:270]: It should not convert them at all. I guess maybe it wrote the ticket incorreclty.

**Review by spocke — CHANGES_REQUESTED** [review]

File: modules/tinymce/src/core/test/ts/browser/html/StylesTest.ts (hunk 3 above)
Comment (spocke) [reply modules/tinymce/src/core/test/ts/browser/html/StylesTest.ts:270]: Also I would want a test on the setContent/getContent APIs as well.

File: modules/tinymce/src/core/main/ts/api/html/Styles.ts [hunk 4]
```diff
@@ -265,6 +269,11 @@ const Styles = (settings: StylesSettings = {}, schema?: Schema): Styles => {
               value = value.toLowerCase();
             }
 
+            // Convert transparent to fully transparent black
+            if (value === 'transparent') {
+              value = '#00000000';
+            }
```
Comment (TheSpyder) [inline modules/tinymce/src/core/main/ts/api/html/Styles.ts:275]: But this isn't what we did in TinyMCE 5, we left it as `transparent`.

Can we not just leave it? The conversion to hex is only supposed to apply to `rgb()` colors, not named colors.

Comment (ArvinJ-H) [reply modules/tinymce/src/core/main/ts/api/html/Styles.ts:275]: Yeah, sorry had some confusing yesterday

File: modules/tinymce/src/core/test/ts/browser/html/StylesTest.ts [hunk 5]
```diff
@@ -264,4 +271,16 @@ describe('browser.tinymce.core.html.StylesTest', () => {
     assertStyles(styles, 'color: rgba(1, 2, 3);', 'color: rgba(1, 2, 3);');
     assertStyles(styles, 'color: rgb(1, 2, 3, 0.5);', 'color: rgb(1, 2, 3, 0.5);');
   });
+
+  it('TINY-10916: transparent should not be converted to other format', () => {
+    const styles = Styles();
+    assertStyles(styles, 'color: transparent;', 'color: transparent;');
```
Comment (tiny-ben-tran) [inline modules/tinymce/src/core/test/ts/browser/html/StylesTest.ts:277]: Should add more test cases to cover other properties that accept color values such as `border-color` and `background-color`

**Review by tiny-ben-tran — APPROVED** [review]

File: modules/tinymce/src/core/main/ts/api/html/Styles.ts [hunk 6]
```diff
@@ -266,7 +267,7 @@ const Styles = (settings: StylesSettings = {}, schema?: Schema): Styles => {
             }
 
             // Convert RGB colors to HEX
-            if (!rgbaRegExp.test(value)) {
+            if (!rgbaRegExp.test(value) && value.toLowerCase() !== 'transparent') {
```
Comment (TheSpyder) [inline modules/tinymce/src/core/main/ts/api/html/Styles.ts:270]: I think there's a more fundamental problem here. If you look at the `rgbaRegExp` and this test, it's converting every value that doesn't begin with `rgba`. That's the problem, specifically whitelisting `transparent` won't help.

**Review by TheSpyder — CHANGES_REQUESTED** [review]

File: modules/tinymce/src/core/main/ts/api/html/Styles.ts (hunk 6 above)
Comment (TheSpyder) [reply modules/tinymce/src/core/main/ts/api/html/Styles.ts:270]: If you look back at TinyMCE 5, where we had the behaviour customers want, it was more or less as I described.
https://github.com/tinymce/tinymce/blob/05989b03bf56d98aaf3522630b481616799c4dc8/modules/tinymce/src/core/main/ts/api/html/Styles.ts#L60
https://github.com/tinymce/tinymce/blob/05989b03bf56d98aaf3522630b481616799c4dc8/modules/tinymce/src/core/main/ts/api/html/Styles.ts#L302

Comment (ArvinJ-H) [reply modules/tinymce/src/core/main/ts/api/html/Styles.ts:270]: Yes, the rgba regex need some fix up, I have updated it to match all 4 values.

The actual cause of the issue is this function
```
RgbaColour.fromString(value).each((rgba) => {}

const fromString = (rgbaString: string): Optional<Rgba> => {
  if (rgbaString === 'transparent') {
    return Optional.some(rgbaColour(0, 0, 0, 0));
  }
 }
 ```
 It recognise `transparent` as a valid string, and convert it into `rgba(0,0,0,0)`, that gets passed into this function
 ```
 const rgbaToHexString = (color: string): string =>
  RgbaColour.fromString(color)
    .map(HexColour.fromRgba)
    .map((h) => '#' + h.value)
    .getOr(color);
```
`fromRgba` function converts colour into `rgb` colour (in this case #000000), hence why the colour black.

Comment (TheSpyder) [reply modules/tinymce/src/core/main/ts/api/html/Styles.ts:270]: Good sleuthing! I think the regex issue is not just that it lacked detail, but swapping the condition logic means that _everything_ which isn't `rgba()` is passed for conversion, rather than just `rgb()` colors as intended.

Removing the `transparent` line looks like a much better fix than adding the `transparent` exclusion here. It's been in the code since 2018, so the exact reason will be lost in an old CI server before it was moved to open source. I can only hope tests fail if it's really needed.

But either way, I think swapping back to an `rgb` check here (perhaps using the regex from `RgbaColour.ts`) is a good idea, since that's clearly the intent of the code.

Comment (ArvinJ-H) [reply modules/tinymce/src/core/main/ts/api/html/Styles.ts:270]: According to documentation, all colours except `rgba` are converted to hex, so the if statement on the regex should be valid.
https://www.tiny.cloud/docs/tinymce/latest/migration-from-6x/#force-hex-color

I'll remove the `transparent` line and see if CI reports anything anywhere.

Comment (spocke) [reply modules/tinymce/src/core/main/ts/api/html/Styles.ts]: I guess we need to update the docs there since transparent was not converted into rgba in v6.

Comment (TheSpyder) [reply modules/tinymce/src/core/main/ts/api/html/Styles.ts]: That's documentation written based on the code. The intent, at least from my perspective, was to return to our v5 behaviour.

**Review by ltrouton — APPROVED** [review]

File: modules/tinymce/src/core/main/ts/api/html/Styles.ts [hunk 7]
```diff
@@ -42,7 +42,7 @@ const Styles = (settings: StylesSettings = {}, schema?: Schema): Styles => {
   const urlOrStrRegExp = /(?:url(?:(?:\(\s*\"([^\"]+)\"\s*\))|(?:\(\s*\'([^\']+)\'\s*\))|(?:\(\s*([^)\s]+)\s*\))))|(?:\'([^\']+)\')|(?:\"([^\"]+)\")/gi;
   const styleRegExp = /\s*([^:]+):\s*([^;]+);?/g;
   const trimRightRegExp = /\s+$/;
-  const rgbaRegExp = /rgba *\(/i;
+  const rgbRegExp = /rgb\s*\(\s*([0-9]+)\s*,\s*([0-9]+)\s*,\s*([0-9]+)\s*\)/i;
```
Comment (TheSpyder) [inline modules/tinymce/src/core/main/ts/api/html/Styles.ts:45]: This doesn't belong here. Please make a function in `RgbaColour` to test using the regex it has instead of repeating it here.

- [ ] Remove this regex

**Review by TheSpyder — APPROVED** [review]
I think this will make a lot of people happy 👍

File: modules/acid/src/main/ts/ephox/acid/api/colour/RgbaColour.ts [hunk 8]
```diff
@@ -130,6 +138,7 @@ export {
   fromHsv,
   fromHex,
   fromString,
+  getColorFormat,
```
Comment (spocke) [inline modules/acid/src/main/ts/ephox/acid/api/colour/RgbaColour.ts:141]: Shouldn't acid also have some tests for this or the results of this change?

File: .changes/unreleased/tinymce-TINY-10916-2024-05-15.yaml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+project: tinymce
+kind: Fixed
+body: Fixed a bug where transparent colors were converted to black.
```
Comment (spocke) [inline .changes/unreleased/tinymce-TINY-10916-2024-05-15.yaml:3]: I think since the category is already Fixed it nice if we can avoid using fixed in the changelog message.

```suggestion
body: CSS color values set to `transparent` were incorrectly converted to '#000000`.
```

**Review by spocke — APPROVED** [review]

File: modules/acid/src/main/ts/ephox/acid/api/colour/RgbaColour.ts [hunk 10]
```diff
@@ -102,13 +105,20 @@ const fromStringValues = (red: string, green: string, blue: string, alpha: strin
...
   const b = parseInt(blue, 10);
   const a = parseFloat(alpha);
+
   return rgbaColour(r, g, b, a);
 };
 
-const fromString = (rgbaString: string): Optional<Rgba> => {
-  if (rgbaString === 'transparent') {
-    return Optional.some(rgbaColour(0, 0, 0, 0));
+const getColorFormat = (colorString: string): string => {
```
Comment (TheSpyder) [inline modules/acid/src/main/ts/ephox/acid/api/colour/RgbaColour.ts:112]: This return should've been a sum type, not a `string`.

All I really wanted was some sort of function to run `rgbRegex.test` - not a full "get format" method when we only ever want to check one of the conditions...
